backend/editor/controllers/search_controller.py

```diff
@@ -0,0 +1,374 @@
+
+@dataclass
+class IsFilterSearchTerm(FilterSearchTerm):
+    filter_value: Literal["root"]
+
+    def build_cypher_query(self, _index: int) -> tuple[str, None]:
+        match self.filter_value:
+            case "root":
+                return "NOT (n)-[:is_child_of]->()", None
+            case _:
+                raise ValueError("Invalid filter value")
+
+
+@dataclass
+class LanguageFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+    negated: bool = False
+
+    def build_cypher_query(self, _index: int) -> tuple[str, None]:
+        if self.negated:
+            return f"n.tags_ids_{self.filter_value} IS NULL", None
+        else:
+            return f"n.tags_ids_{self.filter_value} IS NOT NULL", None
+
+
+@dataclass
+class ParentFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)<-[:is_child_of]-(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class ChildFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)-[:is_child_of]->(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class AncestorFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)<-[:is_child_of*]-(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class DescendantFilterSearchTerm(FilterSearchTerm):
+    filter_value: str
+
+    def build_cypher_query(self, index: int) -> tuple[str, str]:
+        return (
+            "(n)-[:is_child_of*]->(:ENTRY {id: $" + get_query_param_name(index) + "})",
+            self.filter_value,
+        )
+
+
+@dataclass
+class Query:
+    project_id: str
+    name_search_terms: list[str]
+    filter_search_terms: list[FilterSearchTerm]
+
+
+def split_query_into_search_terms(query: str) -> list[str]:
+    """
+    Queries should be split by whitespaces that are not inside quotes
+    """
+    query = query.strip()
+    search_terms = []
+
+    inside_quotes = False
+    term_start = 0
+
+    for term_end in range(len(query)):
+        if query[term_end] == '"':
+            inside_quotes = not inside_quotes
+        # If we are not inside quotes and we encounter a whitespace
+        # we are at the end of the current search term
+        elif query[term_end] == " " and not inside_quotes:
+            search_term = query[term_start:term_end]
+            search_terms.append(search_term)
+            term_start = term_end + 1
+
+    search_terms.append(query[term_start:])
+
+    return search_terms
+
+
+def parse_filter_search_term(search_term: str) -> FilterSearchTerm | None:
+    if ":" not in search_term:
+        return None
```

Comment:
The philosophy behind it is that I wanted to create something that resembled to a parser combinator: the `parse_filter_search_term` has only one objective and that is to try to derive a `FilterSearchTerm` from a `search_term` string and to fail if it has not managed to do so. So it should not manipulate any other kind of object (like a `SearchTerm(value)`).
Then, there is a question of what to do if the function cannot return a `FilterSearchTerm`: I preferred returning `None` instead of throwing an exception because there would only be a single kind of exception (a `ValidatorError("Unable to parse string as FilterSearchTerm")` exception) and because I do not really like exceptions in Python (there is no type checking, i.e. the parent function is not required to handle the exception, and the exception is not treated as a value - contrary to Rust for example). I do understand that the second point may be more contagious than the first one.

Reply:
On exceptions: Really we are in Python, we have to use exceptions (being too far from language culture does not help contribution… you can do it on personal project, or when a closed team strongly agree on a thing, for an open source project it's better to stay in well known patterns).

I'm ok to return None in non contentious patterns (although in this case why not return the string, which can be easily detected and enable some eventual chaining).

But I would like error handling (but I develop that in [the other comment](https://github.com/openfoodfacts/taxonomy-editor/pull/438#discussion_r1524921094))